api/dbadapters/presto.ts

```diff
@@ -0,0 +1,259 @@
+import * as Presto from "presto-client";
+import * as PromisePool from "promise-pool-executor";
+
+import { Credentials } from "df/api/commands/credentials";
+import { IDbAdapter, IDbClient, IExecutionResult } from "df/api/dbadapters/index";
+import { LimitedResultSet } from "df/api/utils/results";
+import { flatten } from "df/common/arrays/arrays";
+import { collectEvaluationQueries, QueryOrAction } from "df/core/adapters";
+import { dataform } from "df/protos/ts";
+
+interface IPrestoExecutionResult {
+  columns?: Presto.IPrestoClientColumnMetaData[];
+  data?: Presto.PrestoClientColumnDatum[];
+  queryId?: string;
+  stats?: Presto.IPrestoClientStats;
+}
+
+// TODO: Move this to somewhere both the API and CLI can use?
+function resolveTarget(target: dataform.ITarget) {
+  return `${target.database ? `${target.database}.` : ""}${
+    target.schema ? `${target.schema}.` : ""
+  }${target.name}`;
+}
+
+export class PrestoDbAdapter implements IDbAdapter {
+  public static async create(credentials: Credentials, options?: { concurrencyLimit?: number }) {
+    return new PrestoDbAdapter(credentials, options);
+  }
+
+  private client: Presto.Client;
+
+  private pool: PromisePool.PromisePoolExecutor;
+
+  private constructor(credentials: Credentials, options?: { concurrencyLimit?: number }) {
+    this.client = new Presto.Client(credentials as Presto.IPrestoClientOptions);
+    this.pool = new PromisePool.PromisePoolExecutor({
+      concurrencyLimit: options?.concurrencyLimit || 10,
+      frequencyWindow: 1000,
+      frequencyLimit: 10
+    });
+  }
+
+  public async execute(
+    statement: string,
+    // TODO: These execute options should actually allow any of Presto.IPrestoClientExecuteOptions.
+    options: {
+      rowLimit?: number;
+      byteLimit?: number;
+    } = { rowLimit: 1000, byteLimit: 1024 * 1024 }
+  ): Promise<IExecutionResult> {
+    let isCancelled = false;
+    const prestoResult: IPrestoExecutionResult = await this.pool
+      .addSingleTask({
+        generator: () =>
+          new Promise((resolve, reject) => {
+            const result: IPrestoExecutionResult = { data: [] };
+            const allRows = new LimitedResultSet({
+              rowLimit: options.rowLimit,
+              byteLimit: options.byteLimit
+            });
+            const cancelIfError = (error: Presto.IPrestoClientError) => {
+              if (!!error) {
+                reject(error);
+                return (isCancelled = true);
+              }
+            };
+            this.client.execute({
+              query: statement,
+              cancel: () => {
+                // The presto client cancels executions if this returns true.
+                return isCancelled;
+              },
+              state: (error, queryId, stats) => {
+                if (cancelIfError(error)) {
+                  return;
+                }
+                result.queryId = queryId;
```

Comment:
while this stuff may be useful in the future, i'm wondering if it's useful right now to keep `IPrestoExecutionResult` around at all. it's a lot of code to keep around when we immediately throw away all the non-`data` fields.

Reply:
Sure, let's just ignore the other stuff for now